azurerm/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go

```diff
@@ -0,0 +1,244 @@
+	serverKeyName := ""
+	serverKeyType := sql.ServiceManaged
+
+	keyVaultKeyId := strings.TrimSpace(d.Get("key_vault_key_id").(string))
+
+	// If it has content, then we assume it's a key vault key id
+	if keyVaultKeyId != "" {
+		// Update the server key type to AKV
+		serverKeyType = sql.AzureKeyVault
+
+		// Set the SQL Server Key properties
+		serverKeyProperties := sql.ServerKeyProperties{
+			ServerKeyType: serverKeyType,
+			URI:           &keyVaultKeyId,
+		}
+		serverKey.ServerKeyProperties = &serverKeyProperties
+
+		// Set the encryption protector properties
+		keyId, err := keyVaultParser.ParseNestedItemID(keyVaultKeyId)
+
+		if err != nil {
+			return fmt.Errorf("Unable to parse key: %q: %+v", keyVaultKeyId, err)
+		}
+
+		// Make sure it's a key, if not, throw an error
+		if keyId.NestedItemType == "keys" {
+			keyName := keyId.Name
+			keyVersion := keyId.Version
+
+			// Extract the vault name from the keyvault base url
+			idURL, err := url.ParseRequestURI(keyId.KeyVaultBaseUrl)
+
+			if err != nil {
+				return fmt.Errorf("Unable to parse key vault hostname: %s", keyId.KeyVaultBaseUrl)
+			}
+
+			hostParts := strings.Split(idURL.Host, ".")
+			vaultName := hostParts[0]
+
+			// Create the key path for the Encryption Protector. Format is: {vaultname}_{key}_{key_version}
+			serverKeyName = fmt.Sprintf("%s_%s_%s", vaultName, keyName, keyVersion)
+		} else {
+			return fmt.Errorf("Key vault key id must be a reference to a key, but got: %s", keyId.NestedItemType)
+		}
+	}
+
+	// Service managed doesn't require a key name
+	encryptionProtectorProperties := sql.EncryptionProtectorProperties{
+		ServerKeyType: serverKeyType,
+		ServerKeyName: &serverKeyName,
+	}
+
+	// Only create a server key if the properties have been set
+	if serverKey.ServerKeyProperties != nil {
+		// Create a key on the server
+		futureServers, err := serverKeysClient.CreateOrUpdate(ctx, serverId.ResourceGroup, serverId.Name, serverKeyName, serverKey)
+		if err != nil {
+			return err
+		}
+
+		if err = futureServers.WaitForCompletionRef(ctx, serverKeysClient.Client); err != nil {
+			return err
+		}
+	}
+
+	encryptionProtectorObject := sql.EncryptionProtector{
+		EncryptionProtectorProperties: &encryptionProtectorProperties,
+	}
+
+	futureEncryptionProtector, err := encryptionProtectorClient.CreateOrUpdate(ctx, serverId.ResourceGroup, serverId.Name, encryptionProtectorObject)
+	if err != nil {
+		return err
+	}
+
+	if err = futureEncryptionProtector.WaitForCompletionRef(ctx, encryptionProtectorClient.Client); err != nil {
+		return fmt.Errorf("waiting on create/update future for Encryption Protector on server %q (Resource Group %q): %+v", serverId.Name, serverId.ResourceGroup, err)
+	}
+
+	// Encryption protector always uses "current" for the name
+	id := parse.NewEncryptionProtectorID(serverId.SubscriptionId, serverId.ResourceGroup, serverId.Name, "current")
```

Comment:
Also here, and we can use the string representation of the ID for convenience/consistency :)

```suggestion
	// Encryption protector always uses "current" for the name
	id := parse.NewEncryptionProtectorID(serverId.SubscriptionId, serverId.ResourceGroup, serverId.Name, "current")

	futureEncryptionProtector, err := encryptionProtectorClient.CreateOrUpdate(ctx, id.ResourceGroup, id.ServerName, encryptionProtectorObject)
	if err != nil {
		return fmt.Errorf("creating/updating %s: %+v", id, err)
	}

	if err = futureEncryptionProtector.WaitForCompletionRef(ctx, encryptionProtectorClient.Client); err != nil {
		return fmt.Errorf("waiting on create/update future for %s: %+v", id, err)
	}

```

Reply:
I assume you mean serverId.ID() instead of id?